codeforamerica/vita-min
# FYST 519 Create role and permissions for NJ in The Hub

## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/FYST-519
## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!
## What was done?
- Adds a "FYST Admin Tools" section to the nav bar of the hub, splitting out the previous "State File Actions" entries in the Admin Tools 
- Creates a new `StateFileNjStaffRole` role, and grant it permission to manage NJ EfileErrors, EfileSubmissions and FaqCategories/FaqItems
- Restrict non-NJ staff from managing NJ EfileErrors, EfileSubmissions or FaqCategories/FaqItems (this last one I'm unsure about, as it wasn't explicitly in the story requirements, so I may change it back)
- Changes some queries to either start with cancancan's loaded resource, or use `accessible_by` if they are "manual" queries that don't start with cancancan's loaded resource
- Use `authorize!` in a few more places to ensure that the currently signed-in user can take certain actions
- Change FaqCategories controller to load and show all our active states automatically, instead of listing them out manually
- Added an NJ FAQ Category product type

Assorted other:
- reduces our usage of explicit before actions like :require_state_file and :require_admin in favor of more specific usages of cancancan's abilities
- fixes some bugs with the FAQ categories and items forms 
  - error messages wouldn't show up in the flash bar at the top
  - on submission of `/new` with validation errors, the product type would disappear and the form would become unusable
  - on submission of `/edit` with validation errors, the product type would switch from the correct state to gyr
- changed a few constraints in the ability.rb file to hopefully be more performant
## How to test?
1. Go to the hub on this heroku deployment
2. Sign in as an admin
3. Invite another email you control as a New Jersey Staff role
4. Accept that invite, go to the hub and sign in, then click FYST Admin Tools and click around to see if things work like they're supposed to
## Screenshots (for visual changes)
![image](https://github.com/user-attachments/assets/d052b6dd-fa17-4960-a266-e9c42f9340fd)
![image](https://github.com/user-attachments/assets/cde6ab3d-d95f-4f8c-9b60-33297461b20c)
![image](https://github.com/user-attachments/assets/28919edf-c863-4b08-8b47-9077701222e7)
